### PR TITLE
Opengl full

### PR DESCRIPTION
**Pull request by devel0**
## What does the pull request do?

Full GL api as from https://github.com/KhronosGroup/OpenGL-Registry/blob/master/xml/gl.xml with documentation from https://github.com/KhronosGroup/OpenGL-Refpages/tree/master/gl4.

## What is the current behavior?

in [GlInterface](https://github.com/AvaloniaUI/Avalonia/blob/06d112e6f117b3dc98199e7f1f130e891052a58d/src/Avalonia.OpenGL/GlInterface.cs) currently there are only some of gl api available without intellisense documentation neither for parameters due to the property get instead of method declaration.

## How was the solution implemented (if it's not obvious)?

Autogenerated using tool from https://github.com/devel0/netcore-sci/tree/d23959d0fba177382cd23241a6cc51a3ca3356c3/tools/gl-csharp-bindings-generator.

## Checklist

- [ ] Added unit tests (if possible)?
- [X ] Added XML documentation to any related classes?
- [ ] Consider submitting a PR to https://github.com/AvaloniaUI/Avaloniaui.net with user documentation

## Breaking changes

- following gl.xml spec some int var should declared as uint
- most methods have enum group

```csharp
var oneArr = new int[1];
gl.GenFramebuffers(1, oneArr);
_fb = oneArr[0];
gl.BindFramebuffer(GL_FRAMEBUFFER, _fb);
```

becomes

```csharp
var oneArr = new uint[1];
gl.GenFramebuffers(1, oneArr);
_fb = oneArr[0];                    
gl.BindFramebuffer(FramebufferTarget.GL_FRAMEBUFFER, _fb);
```

## Fixed issues

none

### CODE REVIEW

**Comment by dnfadmin** [comment]
[CLA assistant check](https://cla.dotnetfoundation.org/AvaloniaUI/Avalonia?pullRequest=4305) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/AvaloniaUI/Avalonia?pullRequest=4305) before we can accept your contribution.<br/><br/>:x: devel0 [sign now](https://cla.dotnetfoundation.org/AvaloniaUI/Avalonia?pullRequest=4305)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/AvaloniaUI/Avalonia?pullRequest=4305) it.</sub>

**Comment by devel0** [comment]
Another possible better solution would to use [OpenGL.Net](https://github.com/luca-piccioni/OpenGL.Net) but I found the way proposed in this PR more immediate in terms of integration with current code and I haven't all the knowledge required to consider that integration; this PR effectively only extends available api methods, document and type enum for them.
Eventually if useful I can consider to add Attributes to methods describing extension that can retrieved from gl.xml spec file.

**Comment by Gillibald** [comment]
You are changing `int` to `uint` all over the place. I don't think that is really needed and just makes it harder to review this.

File: src/Avalonia.OpenGL/GlBasicInfoInterface.cs [hunk 1]
```diff
@@ -16,9 +16,9 @@ public GlBasicInfoInterface(Func<Utf8Buffer, IntPtr> nativeGetProcAddress) : bas
         {
         }
         
-        public delegate void GlGetIntegerv(int name, out int rv);
-        public delegate IntPtr GlGetString(int v);
-        public delegate IntPtr GlGetStringi(int v, int v1);
+        public delegate void bGlGetIntegerv(uint name, out uint rv);
```
Comment (Gillibald) [inline src/Avalonia.OpenGL/GlBasicInfoInterface.cs:19]: Please revert these name changes they bring no benefit and just make it harder to review

**Comment by kekekeks** [comment]
I'm sorry, but we can't accept that to be included in the framework itself. 

1) We have to support rendering through both OpenGL ES and desktop GL, so the idea is to have a **minimum** set of APIs that are present everywhere and absolutely required to get things going.

Our built-in OpenGL bindings are **not intended to be used outside of the framework**. The only APIs that should be used by end-user apps are context management (creation, MakeCurrent) and glGetProcAddress. 

The rest should be done via some exiting OpenGL bindings such as [Silk.NET](https://github.com/ultz/silk.net) [OpenGL.Net](https://github.com/luca-piccioni/OpenGL.Net) and [OpenTK](https://github.com/opentk/opentk). They do support working on top of pre-existing OpenGL context with custom glGetProcAddress delegate.

The general plan was to provide integration shims for those bindings, so there would be `OpenTKControl`, `SilkNETControl`, `OpenGLNetControl` working on top of `OpenGLControlBase` and shipped via separate NuGet packages.

File: samples/ControlCatalog/Pages/OpenGlPage.xaml.cs [hunk 2]
```diff
@@ -362,40 +367,44 @@ protected override unsafe void OnOpenGlRender(GlInterface gl, int fb)
...
         class GlExtrasInterface : GlInterfaceBase<GlInterface.GlContextInfo>
         {
             public GlExtrasInterface(GlInterface gl) : base(gl.GetProcAddress, gl.ContextInfo)
             {
             }
             
-            public delegate void GlDeleteVertexArrays(int count, int[] buffers);
-            [GlMinVersionEntryPoint("glDeleteVertexArrays", 3,0)]
-            [GlExtensionEntryPoint("glDeleteVertexArraysOES", "GL_OES_vertex_array_object")]
+            public delegate void GlDeleteVertexArrays(int count, uint[] buffers);
```
Comment (kekekeks) [inline samples/ControlCatalog/Pages/OpenGlPage.xaml.cs:384]: e. g. this change just breaks the code on OSX.

File: src/Avalonia.Native/Avalonia.Native.csproj [hunk 3]
```diff
@@ -2,7 +2,7 @@
 
   <PropertyGroup>
     <IsPackable>false</IsPackable>
-    <IsPackable Condition="'$([MSBuild]::IsOSPlatform(OSX))' == 'True'">true</IsPackable>
+    <IsPackable Condition="'$([MSBuild]::IsOSPlatform(OSX))' == 'True' Or '$([MSBuild]::IsOSPlatform(Linux))' == 'True'">true</IsPackable>
```
Comment (kekekeks) [inline src/Avalonia.Native/Avalonia.Native.csproj:5]: No, it's not, in fact, packable on Linux

**Comment by devel0** [comment]
@kekekeks Thanks for clarification, I totally agree with you.

I would be useful to specify in the [roadmap](https://github.com/AvaloniaUI/Avalonia/issues/2239) what you said here about OpenTKControl, SilkNETControl, OpenGLNetControl just to allow anyone to know directions.

Consider that I wrote extension proposed in PR just for my need because I'm new to 3d and I got some trouble without having intellisense for example for a function like glPolygonMode etc, but when I did these for my need I thought that opening a PR  could helpful to clarify the question about the infrastructure you described which I eager to see in operation.

Thanks for the work you are forwarding about 3d control availability for the AvaloniaUI.

**Comment by devel0** [comment]
I wrote a PR #4310 that should fit with needs and its very essential,
Hopefully could fit with your requirements to avoid incompatibilities.

**Pull request closed without merging**